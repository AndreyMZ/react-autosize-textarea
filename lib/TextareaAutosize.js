'use strict';

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var React = require('react'),
    autosize = require('autosize');

var TextareaAutosize = React.createClass({
  displayName: 'TextareaAutosize',

  propTypes: {
    onResize: React.PropTypes.func
  },

  getDefaultProps: function getDefaultProps() {
    return {
      rows: 1
    };
  },

  componentDidMount: function componentDidMount() {
    autosize(this.refs.textarea.getDOMNode());
    if (this.props.onResize) {
      this.refs.textarea.getDOMNode().addEventListener('autosize:resized', this.props.onResize);
    }
  },

  componentWillUnmount: function componentWillUnmount() {
    if (this.props.onResize) {
      this.refs.textarea.getDOMNode().removeEventListener('autosize:resized');
    }
  },

  render: function render() {
    return React.createElement(
      'textarea',
      _extends({}, this.props, { ref: 'textarea' }),
      this.props.children
    );
  }

});

module.exports = TextareaAutosize;