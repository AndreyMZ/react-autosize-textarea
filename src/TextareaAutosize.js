'use strict';

const React = require('react'),
  autosize = require('autosize');

const TextareaAutosize = React.createClass({

  propTypes: {
    onResize: React.PropTypes.func
  },

  getDefaultProps() {
    return {
      rows: 1
    };
  },

  componentDidMount() {
    autosize(this.refs.textarea.getDOMNode());
    if (this.props.onResize) {
      this.refs.textarea.getDOMNode().addEventListener('autosize:resized', this.props.onResize);
    }
  },

  componentWillUnmount() {
    if (this.props.onResize) {
      this.refs.textarea.getDOMNode().removeEventListener('autosize:resized');
    }
  },

  render() {
    return (
      <textarea {...this.props} ref='textarea'>
        {this.props.children}
      </textarea>
    );
  }

});

module.exports = TextareaAutosize;