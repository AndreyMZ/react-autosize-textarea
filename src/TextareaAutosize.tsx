import * as React from 'react';
import * as PropTypes from 'prop-types';
import * as autosize from 'autosize';
import * as _getLineHeight from 'line-height';

const getLineHeight = _getLineHeight as (element: HTMLElement) => number | null;

export type TextareaAutosizeRequiredProps = React.TextareaHTMLAttributes<HTMLTextAreaElement> & {
  onResize?: (e: React.SyntheticEvent<Event>) => void,
  onChange?: (e: React.SyntheticEvent<HTMLTextAreaElement>) => void,
  // rows is already typed in `React.TextareaHTMLAttributes<HTMLTextAreaElement>`
  maxRows?: number,
  innerRef?: (textarea: HTMLTextAreaElement) => void
}

export type TextareaAutosizeDefaultProps = {
  rows: number
}

export namespace TextareaAutosize {
  export type Props = TextareaAutosizeRequiredProps & Partial<TextareaAutosizeDefaultProps>;
}

export type State = {
  lineHeight: number | null
}

type EventType = 'autosize:update' | 'autosize:destroy' | 'autosize:resized';

const UPDATE: EventType = 'autosize:update';
const DESTROY: EventType = 'autosize:destroy';
const RESIZED: EventType = 'autosize:resized';

/**
 * A light replacement for built-in textarea component
 * which automaticaly adjusts its height to match the content
 */
export default class TextareaAutosize extends React.Component<TextareaAutosize.Props, State> {

  static defaultProps: TextareaAutosizeDefaultProps = {
    rows: 1
  };

  static propTypes: { [key in keyof TextareaAutosize.Props]: PropTypes.Requireable<any> } = {
    rows: PropTypes.number,
    maxRows: PropTypes.number,
    onResize: PropTypes.func,
    innerRef: PropTypes.func
  }

  state = {
    lineHeight: null
  }

  textarea: HTMLTextAreaElement
  currentValue: TextareaAutosize.Props['value']

  componentDidMount() {
    const { onResize, maxRows } = this.props;

    if (typeof maxRows === 'number') {
      this.updateLineHeight();

      // this trick is needed to force "autosize" to activate the scrollbar
      setTimeout(() => autosize(this.textarea));
    } else {
      autosize(this.textarea);
    }

    if (onResize) {
      this.textarea.addEventListener(RESIZED, onResize as any);
    }
  }

  componentWillUnmount() {
    const { onResize } = this.props;
    if (onResize) {
      this.textarea.removeEventListener(RESIZED, onResize as any);
    }
    this.dispatchEvent(DESTROY);
  }

  dispatchEvent = (EVENT_TYPE: EventType) => {
    const event = document.createEvent('Event');
    event.initEvent(EVENT_TYPE, true, false);

    this.textarea.dispatchEvent(event);
  };

  updateLineHeight = () => {
    this.setState({
      lineHeight: getLineHeight(this.textarea)
    });
  }

  onChange = (e: React.SyntheticEvent<HTMLTextAreaElement>) => {
    const { onChange } = this.props;
    this.currentValue = e.currentTarget.value;
    onChange && onChange(e);
  }

  saveDOMNodeRef = (ref: HTMLTextAreaElement) => {
    const { innerRef } = this.props;

    if (innerRef) {
      innerRef(ref);
    }

    this.textarea = ref;
  }

  getLocals = () => {
    const {
      props: { onResize, maxRows, onChange, style, innerRef, ...props },
      state: { lineHeight },
      saveDOMNodeRef
    } = this;

    const maxHeight = maxRows && lineHeight ? lineHeight * maxRows : null;

    return {
      ...props,
      saveDOMNodeRef,
      style: maxHeight ? { ...style, maxHeight } : style,
      onChange: this.onChange
    };
  }

  render() {
    const { children, saveDOMNodeRef, ...locals } = this.getLocals();
    return (
      <textarea {...locals} ref={saveDOMNodeRef}>
        {children}
      </textarea>
    );
  }

  componentDidUpdate() {
    if (this.props.value !== this.currentValue) {
      this.dispatchEvent(UPDATE);
    }
  }

}
