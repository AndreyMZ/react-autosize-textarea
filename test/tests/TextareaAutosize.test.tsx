import * as React from 'react';
import * as renderer from 'react-test-renderer';
import TextareaAutosize from '../../src';

describe('TextareaAutosize', () => {

  it('renders correctly', () => {
    const component = renderer.create(
      <TextareaAutosize defaultValue='Initial Value' />
    );
    expect(component).toMatchSnapshot();
  });

});
