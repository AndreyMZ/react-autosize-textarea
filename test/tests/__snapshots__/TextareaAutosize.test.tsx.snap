// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextareaAutosize renders correctly 1`] = `
<textarea
  defaultValue="Initial Value"
  onChange={[Function]}
  rows={1}
  style={undefined}
/>
`;

exports[`build build script generates every needed file 1`] = `
Array [
  "TextareaAutosize.d.ts",
  "TextareaAutosize.js",
  "index.d.ts",
  "index.js",
]
`;
